### PR TITLE
GraphScope on 3/3 part 1: add local/dist trainer, and a `Data` class to make the example simpler

### PR DESCRIPTION
**Pull request by sighingnow**
This pull request

- unify the logic of local/dist trainer
- add a `Data` class to gcn and rgcn to make the example code simpler, as discussed
- ~~add a `Ego` based GCN to restore the previous GCN example, and show how `EgoXXXData` works.~~

### CODE REVIEW

File: graphlearn/examples/tf/ego_data.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    self.val_batch_size = val_batch_size
+    self.sampler = sampler
+
+    # train
+    tfg.conf.training = True
+    self.query_train = self.query(self.graph, gl.Mask.TRAIN)
+    self.dataset_train = tfg.Dataset(self.query_train, window=10)
+    self.train_iterator = self.dataset_train.iterator
+    self.train_dict = self.dataset_train.get_data_dict()
+    self.train_embedding = self.model.forward(
```
Comment (Seventeen17) [inline graphlearn/examples/tf/ego_data.py:54]: I think it's better not to encapsulate the model training into the `Data`.

Comment (sighingnow) [reply graphlearn/examples/tf/ego_data.py:54]: Is moving `self.{train,val,test}_embedding` to outside and still keeping other field in `Data` class acceptable?

File: graphlearn/examples/tf/trainer.py [hunk 2]
```diff
@@ -38,141 +39,181 @@
...
-            file_name = 'timeline_' + str(local_step) + '_' + \
-              str(self.task_index) + '.json'
-            save_path = os.path.join(self.ckpt_dir, file_name)
-            writeGFile = tf.gfile.GFile(save_path, mode='w')
-            writeGFile.write(content)
-            writeGFile.flush()
-            writeGFile.close()
-            print("Profiling data save to %s success." % save_path)
+          if self.profiling:
+            outs = self.run_and_profiling(train_ops, local_step)
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply graphlearn/examples/tf/trainer.py:135]: I think move `self.profiling` to `run_and_profiling()` and use `run_and_profiling` directly would look better.

File: graphlearn/examples/tf/ego_data.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+  import tensorflow.compat.v1 as tf
+  tf.disable_v2_behavior()
+except ImportError:
+  import tensorflow as tf
+
+import graphlearn as gl
+import graphlearn.python.nn.tf as tfg
+from graphlearn.python.utils import parse_nbrs_num
+
+class EgoData:
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply graphlearn/examples/tf/ego_data.py:37]: Looks fine, thanks!

File: graphlearn/python/nn/tf/layers/ego_gcn_conv.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  import tensorflow.compat.v1 as tf
+  tf.disable_v2_behavior()
+except ImportError:
+  import tensorflow as tf
+
+from graphlearn.python.nn.tf.layers.ego_layer import EgoConv, EgoLayer
+from graphlearn.python.nn.tf.layers.linear_layer import LinearLayer
+
+
+class EgoGCNConv(EgoConv):
```
Comment (baoleai) [inline graphlearn/python/nn/tf/layers/ego_gcn_conv.py:33]: It seems like EgoRGCNConv, not EgoGCN. You can use EgoSAGEConv with aggr='gcn' as EgoGCNConv.

File: graphlearn/examples/tf/ego_gcn/ego_gcn.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+               num_layers,
+               num_bases=None,
+               num_blocks=None,
+               agg_type="mean",
+               use_bias=False,
+               bn_func=None,
+               act_func=tf.nn.relu,
+               dropout=0.0,
+               **kwargs):
+    """EgoGraph based RGCN.
```
Comment (baoleai) [inline graphlearn/examples/tf/ego_gcn/ego_gcn.py:43]: These args are for RGCN not for GCN.

File: graphlearn/examples/tf/ego_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (baoleai) [reply graphlearn/examples/tf/ego_data.py:37]: It is better to put all sampling and data preprocessing into a `SampleLoader` or `NeighborLoader`.

File: graphlearn/examples/tf/ego_data.py (hunk 1 above)
Comment (baoleai) [reply graphlearn/examples/tf/ego_data.py:54]: Move all model-related data outside.

File: graphlearn/python/nn/tf/layers/ego_gcn_conv.py (hunk 4 above)
Comment (sighingnow) [reply graphlearn/python/nn/tf/layers/ego_gcn_conv.py:33]: Okay. will test that. Thanks!

File: graphlearn/examples/tf/ego_gcn/ego_gcn.py (hunk 5 above)
Comment (sighingnow) [reply graphlearn/examples/tf/ego_gcn/ego_gcn.py:43]: Thanks, will fix.

File: graphlearn/examples/tf/ego_rgcn/train_supervised.py [hunk 6]
```diff
@@ -183,37 +117,25 @@ def main(unused_argv):
                   FLAGS.num_blocks,
                   agg_type=FLAGS.agg_type,
                   dropout=FLAGS.drop_out)
+  # prepare train dataset
+  train_data = EgoRGCNDataLoader(g, gl.Mask.TRAIN, FLAGS.sampler, FLAGS.train_batch_size,
+                                 node_type='i', nbrs_num=nbrs_num, num_relations=FLAGS.num_relations)
+  train_embedding = model.forward(train_data.as_list(), nbrs_num)
+  loss = supervised_loss(train_embedding, train_data['seed'].labels)
```
Comment (baoleai) [inline graphlearn/examples/tf/ego_rgcn/train_supervised.py:124]: The user cannot know 'seed', it is inside 'train_data`, so maybe just use API like `seed().labels` ?

File: graphlearn/examples/tf/ego_data.py (hunk 1 above)
Comment (sighingnow) [reply graphlearn/examples/tf/ego_data.py]: Fixed.

File: graphlearn/examples/tf/ego_data.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sighingnow) [reply graphlearn/examples/tf/ego_data.py]: Fixed.

File: graphlearn/examples/tf/trainer.py [hunk 7]
```diff
@@ -38,141 +39,178 @@
...
+        **checkpoint_args,
+        **kwargs)
 
     def _close_session():
       if self.sess is not None:
         self.sess.close()
     atexit.register(_close_session)
 
-  def train(self, iterator, loss, learning_rate, epochs=10, hooks=[], **kwargs):
+  def run_and_profiling(self, train_ops, local_step):
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply graphlearn/examples/tf/trainer.py]: Fixed. Has been renamed to `run_step`.

File: graphlearn/examples/tf/ego_rgcn/train_supervised.py (hunk 6 above)
Comment (sighingnow) [reply graphlearn/examples/tf/ego_rgcn/train_supervised.py]: I have changed back to `train/test/val` based on the `mask` parameter (keep the previous behaviour) and expose some helpers `train_labels`, `test_labels`, `val_labels` for accessing.

File: graphlearn/examples/tf/ego_data.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    return self._data_dict[key]
+
+  def __getitem__(self, key):
+    return self.data(key)
+
+  def get_egograph(self, key):
+    return self._dataset.get_egograph(key)
+
+  @property
+  def train_ego(self):
```
Comment (baoleai) [inline graphlearn/examples/tf/ego_data.py:76]: Here should keep only a single `ego()` property in which you can `get_egograph` according to `self_mask`.

File: graphlearn/examples/tf/trainer.py [hunk 9]
```diff
@@ -38,141 +39,180 @@
...
 
-class DistTrainer(object):
-  """Class for distributed training and evaluation
+class TFTrainer(object):
+  """Class for local or distributed training and evaluation.
 
   Args:
-    cluster_spec: TensorFlow ClusterSpec.
-    job_name: name of this worker.
     task_index: index of this worker.
```
Comment (baoleai) [inline graphlearn/examples/tf/trainer.py:46]: rm this unused arg.

File: graphlearn/examples/tf/trainer.py [hunk 10]
```diff
@@ -38,141 +39,180 @@
...
+    self.conf.gpu_options.allow_growth = True
+    self.conf.allow_soft_placement = False
     self.sess = None
 
+    # use for distributed training
+    self.sync_barrier = None
+    self.global_step = None
+    self.is_local = None
+
   def context(self):
```
Comment (baoleai) [inline graphlearn/examples/tf/trainer.py:74]: raise NotImplementedError

File: graphlearn/examples/tf/trainer.py [hunk 11]
```diff
@@ -38,141 +39,180 @@
...
+  def test(self, iterator, test_acc, hooks=[], **kwargs):
+    with self.context():
+      self.global_step = tf.train.get_or_create_global_step()
+      if self.sess is None:
+        self.init_session(hooks=hooks)
+
+      print('Start testing ...')
+      total_test_acc = []
+      local_step = 0
+      last_local_step = 0
```
Comment (baoleai) [inline graphlearn/examples/tf/trainer.py:177]: I think the `LocalTrainer` can also use `global_step`?

File: graphlearn/examples/tf/ego_data.py (hunk 8 above)
Comment (baoleai) [reply graphlearn/examples/tf/ego_data.py:76]: Here we need `src_ego`, `dst_ego` and `neg_dst_ego` three methods(`neg_dst_ego` for unsupervised model). `train_ego` or `test_ego` is just the case of `src_ego` when mask is Train or Test.

File: graphlearn/examples/tf/ego_data.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
+# Copyright 2021 Alibaba Group Holding Limited. All Rights Reserved.
```
Comment (baoleai) [inline graphlearn/examples/tf/ego_data.py:1]: ego_data.py ->ego_data_loader.py

File: graphlearn/examples/tf/ego_data.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    ''' Alias for `self.get_egograph('dst')`.
+    '''
+    return self.get_egograph('dst')
+
+  @property
+  def labels(self):
+    prefix = ('train', 'test', 'val')[self._mask.value - 1]
+    return self._data_dict[prefix].labels
+
+  def as_list(self):
```
Comment (baoleai) [inline graphlearn/examples/tf/ego_data.py:104]: change to `x_list` which means the input node feature(processed) list.

Comment (sighingnow) [reply graphlearn/examples/tf/ego_data.py]: Fixed.

File: graphlearn/examples/tf/ego_data.py (hunk 12 above)
Comment (sighingnow) [reply graphlearn/examples/tf/ego_data.py]: Done, renamed `ego_rgcn_data_loader.py` and `ego_sage_data_loader.py` as well.

File: graphlearn/examples/tf/trainer.py [hunk 14]
```diff
@@ -38,141 +39,180 @@
...
+        **checkpoint_args,
+        **kwargs)
 
     def _close_session():
       if self.sess is not None:
         self.sess.close()
     atexit.register(_close_session)
 
-  def train(self, iterator, loss, learning_rate, epochs=10, hooks=[], **kwargs):
+  def run_step(self, train_ops, local_step):
```
[the remark this answers is not in the document]
Comment (sighingnow) [reply graphlearn/examples/tf/trainer.py:102]: Fixed.

File: graphlearn/examples/tf/trainer.py (hunk 10 above)
Comment (sighingnow) [reply graphlearn/examples/tf/trainer.py:74]: Done.

File: graphlearn/examples/tf/trainer.py (hunk 9 above)
Comment (sighingnow) [reply graphlearn/examples/tf/trainer.py]: Docstring has been revised in trainer.py.

File: graphlearn/examples/tf/ego_data.py (hunk 8 above)
Comment (sighingnow) [reply graphlearn/examples/tf/ego_data.py]: > We should not put train/test/val queries all together for supporting user who only want to run save embedding phase.

Fixed.

File: graphlearn/examples/tf/ego_data_loader.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+      return self.get_egograph('src')
+    else:
+      prefix = ('train', 'test', 'val')[self._mask.value - 1]
+      return self.get_egograph(prefix)
+
+  @property
+  def dst_ego(self):
+    ''' Alias for `self.get_egograph('dst')`.
+    '''
+    return self.get_egograph('dst')
```
Comment (Seventeen17) [inline graphlearn/examples/tf/ego_data_loader.py:89]: It seems that not all the queries in sub class contains 'src', 'dst' and 'neg_dst'

Comment (sighingnow) [reply graphlearn/examples/tf/ego_data_loader.py:89]: They only be called when needed, otherwise we would need user to hard code `get_egograph("src"), ...` in their `train_(un)supervised.py`.

File: graphlearn/examples/tf/ego_data_loader.py [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+    return self.data(key)
+
+  def get_egograph(self, key):
+    return self._dataset.get_egograph(key)
+
+  @property
+  def src_ego(self):
+    ''' Alias for `self.get_egograph('src')`.
+    '''
+    if self._mask is None:
```
Comment (baoleai) [inline graphlearn/examples/tf/ego_data_loader.py:79]: The base class should not provide default implementation for `src_ego`, `dst_ego`, `neg_dst_ego`, because 'src' or 'dst' should only be valid when query use it in derived class. 
You can just raise NotImplementedError here.

File: graphlearn/examples/tf/ego_data_loader.py [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    '''
+    return self.get_egograph('dst')
+
+  @property
+  def neg_dst_ego(self):
+    ''' Alias for `self.get_egograph('neg_dst')`.
+    '''
+    return self.get_egograph('neg_dst')
+
+  @property
```
Comment (baoleai) [inline graphlearn/examples/tf/ego_data_loader.py:97]: `labels` , `x_list` and `_format` these interfaces are not so common. It is best to implement them in the required subclasses such as EgoRGCNDataLoader, not in the base class.

**Review by baoleai — APPROVED** [review]

**Review by Seventeen17 — APPROVED** [review]
